File: src/YowThi.Erp.Application/Party/OutsourcedVendorMaster.cs
using YowThi.Erp.Application.Common.Errors;
using YowThi.Erp.Application.Common.Idempotency;
using YowThi.Erp.Application.Common.Identity;
using YowThi.Erp.Application.Common.Results;

namespace YowThi.Erp.Application.Party;

public sealed record CreateOutsourcedVendorCommand(
    string? NameZhTw,
    string? NameThTh,
    string? BankName,
    string? BankAccount,
    string? Phone,
    string? Address,
    bool Active,
    string? Code = null);

public sealed record CreateOutsourcedVendorExecution(
    CommandId CommandId,
    CommandRequestHash RequestHash,
    ActorAccountId ActorAccountId,
    CreateOutsourcedVendorCommand Command);

public sealed record UpdateOutsourcedVendorCommand(
    Guid OutsourcedVendorId,
    long ExpectedRowVersion,
    string? NameZhTw,
    string? NameThTh,
    string? BankName,
    string? BankAccount,
    string? Phone,
    string? Address,
    bool Active,
    string? Code = null);

public sealed record UpdateOutsourcedVendorExecution(
    CommandId CommandId,
    CommandRequestHash RequestHash,
    ActorAccountId ActorAccountId,
    UpdateOutsourcedVendorCommand Command);

public sealed record OutsourcedVendorMasterWriteResult(
    Guid OutsourcedVendorId,
    long RowVersion);

public interface IOutsourcedVendorMasterExecutor
{
    ValueTask<ApplicationResult<OutsourcedVendorMasterWriteResult>> CreateAsync(
        CreateOutsourcedVendorExecution execution,
        CancellationToken cancellationToken);

    ValueTask<ApplicationResult<OutsourcedVendorMasterWriteResult>> UpdateAsync(
        UpdateOutsourcedVendorExecution execution,
        CancellationToken cancellationToken);
}

public enum OutsourcedVendorMasterStatusFilter
{
    All,
    Active,
    Inactive,
    Deleted,
}

public sealed record OutsourcedVendorMasterQuery(
    string? Search,
    OutsourcedVendorMasterStatusFilter Status,
    int Offset,
    int Limit);

public sealed record OutsourcedVendorMasterItem(
    Guid Id,
    string? Code,
    string? NameZhTw,
    string? NameThTh,
    string? BankName,
    string? BankAccount,
    string? Phone,
    string? Address,
    bool Active,
    long RowVersion,
    DateTimeOffset CreatedAt,
    DateTimeOffset? DeletedAt);

public sealed record OutsourcedVendorMasterPage(
    IReadOnlyList<OutsourcedVendorMasterItem> Items,
    int? NextOffset);

public interface IOutsourcedVendorMasterReader
{
    ValueTask<OutsourcedVendorMasterPage> GetAsync(
        OutsourcedVendorMasterQuery query,
        CancellationToken cancellationToken);
}

public static class OutsourcedVendorMasterErrorCodes
{
    public const string InvalidInput = "party.outsourced-vendor.invalid-input";
    public const string OutsourcedVendorNotFound = "party.outsourced-vendor.not-found";
    public const string OutsourcedVendorDeleted = "party.outsourced-vendor.deleted";
    public const string StaleRowVersion = "party.outsourced-vendor.stale-row-version";
    public const string IdempotencyKeyReused = "party.outsourced-vendor.idempotency-key-reused";
}

public static class OutsourcedVendorMasterValidation
{
    public static ApplicationError? Validate(CreateOutsourcedVendorCommand command)
    {
        ArgumentNullException.ThrowIfNull(command);
        return HasVisibleName(command.NameZhTw, command.NameThTh)
            ? null
            : Invalid();
    }

    public static ApplicationError? Validate(UpdateOutsourcedVendorCommand command)
    {
        ArgumentNullException.ThrowIfNull(command);
        return command.OutsourcedVendorId != Guid.Empty
            && command.ExpectedRowVersion >= 1
            && HasVisibleName(command.NameZhTw, command.NameThTh)
                ? null
                : Invalid();
    }

    private static bool HasVisibleName(string? nameZhTw, string? nameThTh) =>
        !string.IsNullOrWhiteSpace(nameZhTw) || !string.IsNullOrWhiteSpace(nameThTh);

    private static ApplicationError Invalid() =>
        ApplicationError.Create(
            ApplicationErrorKind.Validation,
            OutsourcedVendorMasterErrorCodes.InvalidInput);
}

File: src/YowThi.Erp.Web/src/features/party/outsourcedVendorMaster.ts
import { getApiJson, postApiCommand } from '../../app/api/apiTransport';
import type { OperationalLocale } from '../../app/i18n/locale';

export { ApiProblemError } from '../../app/api/apiTransport';

export type OutsourcedVendorMasterStatus = 'all' | 'active' | 'inactive' | 'deleted';

export interface OutsourcedVendorMasterItem {
  id: string;
  code: string | null;
  nameZhTw: string | null;
  nameThTh: string | null;
  bankName: string | null;
  bankAccount: string | null;
  phone: string | null;
  address: string | null;
  active: boolean;
  rowVersion: number;
  createdAt: string;
  deletedAt: string | null;
}

export interface OutsourcedVendorMasterPage {
  items: OutsourcedVendorMasterItem[];
  nextOffset: number | null;
}

export interface OutsourcedVendorDraftRequest {
  code: string | null;
  nameZhTw: string | null;
  nameThTh: string | null;
  bankName: string | null;
  bankAccount: string | null;
  phone: string | null;
  address: string | null;
  active: boolean;
}

export type CreateOutsourcedVendorRequest = OutsourcedVendorDraftRequest;

export interface UpdateOutsourcedVendorRequest extends OutsourcedVendorDraftRequest {
  expectedRowVersion: number;
}

export interface OutsourcedVendorWriteResult {
  outsourcedVendorId: string;
  rowVersion: number;
}

interface OutsourcedVendorQuery {
  locale: OperationalLocale;
  search?: string;
  status?: OutsourcedVendorMasterStatus;
  offset?: number;
  limit?: number;
  signal?: AbortSignal;
}

interface OutsourcedVendorCommandOptions {
  locale: OperationalLocale;
  idempotencyKey: string;
  signal?: AbortSignal;
}

export function listOutsourcedVendors(query: OutsourcedVendorQuery): Promise<OutsourcedVendorMasterPage> {
  const parameters = new URLSearchParams();
  const search = query.search?.trim();
  if (search) parameters.set('search', search);
  if (query.status) parameters.set('status', query.status);
  if (query.offset !== undefined) parameters.set('offset', String(query.offset));
  if (query.limit !== undefined) parameters.set('limit', String(query.limit));
  const suffix = parameters.size === 0 ? '' : `?${parameters.toString()}`;
  return getApiJson<OutsourcedVendorMasterPage>(`/api/v1/party/outsourced-vendors${suffix}`, {
    locale: query.locale,
    signal: query.signal,
  });
}

export function createOutsourcedVendor(
  request: CreateOutsourcedVendorRequest,
  options: OutsourcedVendorCommandOptions,
): Promise<OutsourcedVendorWriteResult> {
  return postApiCommand<CreateOutsourcedVendorRequest, OutsourcedVendorWriteResult>(
    '/api/v1/party/outsourced-vendors',
    request,
    options,
  );
}

export function updateOutsourcedVendor(
  outsourcedVendorId: string,
  request: UpdateOutsourcedVendorRequest,
  options: OutsourcedVendorCommandOptions,
): Promise<OutsourcedVendorWriteResult> {
  return postApiCommand<UpdateOutsourcedVendorRequest, OutsourcedVendorWriteResult>(
    `/api/v1/party/outsourced-vendors/${outsourcedVendorId}/update`,
    request,
    options,
  );
}
